service_config.md

```diff
@@ -0,0 +1,194 @@
+                     +----------------------------- |                                                 (1)                                |
+                     |                              | [service_user]                                                                     |
+                     |                              | password = {{.Password}}                                                           |
+                     |                              +------------------------------------------------------------------------------------+
+                     |
++----------------------------------------+      +----------------------------------------------------------------------------------------+
+| apiVersion: v1                   (2)   |      |  “config_files”: [        (3)                                                          |
+| metadata:                              |      |                  ...                                                                   |
+|   name: 01-<service>-deployment-secret |      |      {                                                                                 |
+|   namespace: openstack                 | ===> |        "source": "/var/lib/config-data/deployment/01-deployment-<service>-secret.conf",|
+| stringData:                            |      |        "dest": "/etc/<service>/<service>.conf.d/01-deployment-<service>-secret.conf",  |
+|   01-deployment-<service>.conf: |      |      |        "owner": "<service>",                                                           |
+|        <data>                          |      |        "perm": "0600"                                                                  |
++----------------------------------------+      |      },                                                                                |
+                                                +----------------------------------------------------------------------------------------+
+```
+
+As the diagram above depicts, a given operator is supposed to implement the
+logic to build and reconcile the `Secret` when the template has been processed.
+The `Secret`, mounted to the resulting service Pod, in the very last step of
+this process is synced by kolla to the destination directory: instead of having
+a bash script (e.g. `init.sh`) doing the `chown` on the destination folder,
+we'll rely (where possible) on kolla that allows setting ownership and
+permissions, as well as copying optional files in case additional configuration
+is provided.
+
+
+## Render multiple secrets: “CustomServiceConfigSecret” interface
+
+Currently the `OpenStack` storage operators expose the
+`customConfigServiceSecret` parameter, which provides a mechanism for
+specifying service configs via Secrets. Instead of specifying sensitive config
+data directly in the `customServiceConfig`, a cloud admin can place sensitive
+data in a `Secret`, and reference the secret by name in the service's
```

Comment:
What will be the expected structure of the Secret? 
* a) directly storing key value pairs in the secret that are interpreted as oslo config options and their values? If so how we model oslo config sections
* b) a Secret expects to hold a single key value pair. Where the key will be fixed (like `snipet`) and the value is a string that holds a oslo config file snippet.
* c) a Secret can have many key value pairs. Each key is a name of a config file and the value is string that holds an solo config file snippet. How will the operator order the different files from a single Secret in the conf.d directory?

What will happen if the user specifies more than one secret name in `customServiceConfigSecrets`? Will the operator renders them as separate files. Therefore not one `04-secrets.conf` will be created but multiple files. If so I suggest to name the files based on `<index>-<name-of-the-secret>.conf` and use a running index that ensures the order of the secrets in `customServiceConfigSecrets` is the same as the order of the files in the conf.d dir.

Reply:
Looking more at the example below I think I can answer some of my questions.

The `customServiceConfigSecrets` field is a list of string field, where each string is the name of a k8s Secret resource.

Each of these Secrets can have arbitrary amount of arbitrarily named a keys (e.g. snipet, foo, bar) the value at these keys are olso config file snippets. The the operator will concatenate all the values from all the secrets to get 04_secret.conf. E.g. something like:
```golang
var config4 string = ""
for _, secret := range instance.Spec.CustomServiceConfigSecrets {
    for _, v := range secret.Data {
        config4 += v + "\n"
    }
}
```

The oslo config allows repeating a configuration key. repetition has different meaning depending on the type of the oslo config option. The repetition can mean override the previous value, or can mean add the new value to the list of previous values of the config option. Due to the former case we need to be very explicit about the order of the concatenation as the last value of the repeated config option will be used by the service. In the above pseudo-code the secretes are iterated in the order of the list in `customServiceConfigSecrets` so that can be documented to the end user. But the iteration of the secret.Data keys are undefined. (golang map iteration order is random). Even if we sort the keys in the secret.Data we need to be very explicit in the documentation that even though we are not use the keys from the Secret directly, renaming a key could result in a different iteration order and therefore a different service config.

I suggest to be more explicit about the order and use the keys from the secret.Data as the name of the files we actually produce for the service. Then we can point to the oslo config doc that states the order in which the files are loaded. (This means instead of 04_secret.conf we might generate arbitrary amount of fiels in conf.d)

Reply:
i was orignally expectign the customServiceConfigSecret to be a single secret name with a single hardcoded key containing a string which is the content of the file.

Reply:
1. This is also potentially overriding keys between secrets.
```
mergedData = append(mergedData, v...)
```
E.g. if there are two secretes in `customServiceConfigSecrets` both having a single key `foo` with different snippets as values then this code will simply drop the value of the `foo` key from the first secret as the second will override it.

2. Also this code does not concatenate snippets within the resulting single secret into a single key that can be mapped to 04_secret.conf.

Reply:
Sorry disregard my comment. I missunderstood the snipet.

Reply:
So my only remaining comment here is that the concatenation order needs to be well defined as the meaning of resulting config file depends on it. The ` for _, v := range data {}` construct will iterate the map in a random order which is problematic.

Reply:
Overriding: yes -> as I mentioned, it blindly concatenate things: the secrets are created by the human operator, not sure we have too much control over that content;
Concatenate: it works because I tried it, but we can do the `strings()` version +"\n", I'm not against that approach

Reply:
> Overriding: yes -> as I mentioned, it blindly concatenate things: the secrets are created by the human operator, not sure we have too much control over that content;

I see two major option to make this clear for the human operator:
* a) remove ambiguity by not allowing more than one config snippet as input for 04-secrets.conf. This would mean the `customServiceConfigSecret` needs to hold only a single Secret, with a single (probably predefined) keys. Then no concatenation happens so no ordering issue araises.

* b) define a static ordering and document it clearly. `customServiceConfigSecrets` can hold multiple secrets and each secret can have many, (even overlapping) keys. We document that we concatenate the content of the secrets in `customServiceConfigSecrets` by
  * iterating the list of Secrets in the order how they appear the list in `customServiceConfigSecrets`
  * iterating the keys in each Secret in lexicographical order

E.g.:
```
customServiceConfigSecrets:
    foo-secret
    bar-secret
//--
FooSecret
    data:
        pear: |
            an-oslo-conf-option = foo-pear
        apple: |
            an-oslo-conf-option = foo-apple
//--
BarSecret
    data:
        pear: |
            an-oslo-conf-option = bar-pear
        apple: |
            an-oslo-conf-option = bar-apple
```
Will result in the 04-secret.conf file:
```
            an-oslo-conf-option = foo-apple
            an-oslo-conf-option = foo-pear
            an-oslo-conf-option = bar-apple
            an-oslo-conf-option = bar-pear
```
So the lexicographically biggest key of the last Secret from `customServiceConfigSecrets` wins, if mutliple keys define the same oslo config option.

Reply:
option A is my preference by far.
customServiceConfigSecret: myCustomSecret
//--
myCustomSecret:
  StringData:
    config: |
      [section-1]
      key=value
      [meaning]
      life=42

Reply:
right, geven the outlined oslo config specifics, and (deep?)merging complexity (nesting lists to add or replace?, keys ordering, etc.) please specify the case for having multiple secrets referenced , which cannot be covered by a single secret (aside of the 1.5M size limitation, iirc)?

Reply:
Personally I think we should just document how secrets are used here, which is that operators will drop everything within a secret without following an order, so human operators cannot use a single secret to store different things.

For example, they probably don't want to use a single secret to store ALL their cinder backend secrets, otherwise they will be present in all the different cinder-volume pods even if they are not used.

Reply:
@Akrog Thanks for the use case. I think this is a good example to support multiple Secrets. I still don't see a use case to allow multiple snippets per Secret.

Reply:
I don't think it will be used, but I may be missing some use case.

In my opinion the reason to support multiple snippets per secret is simplicity.  Basically we say that we drop everything within a secret instead of explaining that we'll only use the first snippet.

Reply:
for multiple backends, ther could be still a single secret like:
```
kind: Secret
  metadata:
  name: cinder-volume-backends
data:
  backend1.conf: |
  backend2.conf: |
```

Reply:
If I understood correctly the current proposal is to only support whole secrets, not allow using specific keys.

In your example we wouldn't be able to mount "backend1.conf" and "fc_zone_secret", we would have to use everything.

The recommendation for cinder is going to be to have 1 pod per backend (there are many, many reasons why this is a superior option) and in that case mounting the whole secret would make services have more information that they should.  Don't get me wrong, it would work, but it may not be the best in terms of security.

If a customer prefers to go with the multi-backend per cinder-volume pod, then that configuration would be fine.

Reply:
> 



> > Overriding: yes -> as I mentioned, it blindly concatenate things: the secrets are created by the human operator, not sure we have too much control over that content;
> 
> I see two major option to make this clear for the human operator:
> 
>     * a) remove ambiguity by not allowing more than one config snippet as input for 04-secrets.conf. This would mean the `customServiceConfigSecret` needs to hold only a single Secret, with a single (probably predefined) key. Then no concatenation happens so no ordering issue araises.
> 
>     * b) define a static ordering and document it clearly. `customServiceConfigSecrets` can hold multiple secrets and each secret can have many, (even overlapping) keys. We document that we concatenate the content of the secrets in `customServiceConfigSecrets` by
>       
>       * iterating the list of Secrets in the order how they appear the list in `customServiceConfigSecrets`
>       * iterating the keys in each Secret in lexicographical order
> 
> 
> E.g.:
> 
> ```
> customServiceConfigSecrets:
>     foo-secret
>     bar-secret
> //--
> FooSecret
>     data:
>         pear: |
>             an-oslo-conf-option = foo-pear
>         apple: |
>             an-oslo-conf-option = foo-apple
> //--
> BarSecret
>     data:
>         pear: |
>             an-oslo-conf-option = bar-pear
>         apple: |
>             an-oslo-conf-option = bar-apple
> ```
> 
> Will result in the 04-secret.conf file:
> 
> ```
>             an-oslo-conf-option = foo-apple
>             an-oslo-conf-option = foo-pear
>             an-oslo-conf-option = bar-apple
>             an-oslo-conf-option = bar-pear
> ```
> 
> So the lexicographically biggest key of the last Secret from `customServiceConfigSecrets` wins, if mutliple keys define the same oslo config option.

Thanks @gibizer. We don't have control over the secrets created by the human operator, hence I'd like to document that  `customServiceConfigSecrets` can hold multiple secrets (because it's an array) and we strongly recommend to not store multiple snippets in a single secret because it can result in a not predictable configuration.
For the record, I think we could (and we should) do the iteration in lexicographical order (so that the last one win): it's something we had in `TripleO` with regard to the `Heat` parameter processing, but given this is a generic high level document we can leave this as a design choice for each operator and document it in the related section/ (e.g. see [1]).
Do you think it's reasonable?

[1] https://github.com/openstack-k8s-operators/glance-operator/blob/main/docs/dev/design-decisions.md

Reply:
> customServiceConfigSecrets can hold multiple secrets

OK by me as there is a use case for it mentioned above

> we strongly recommend to not store multiple snippets in a single secret

OK. 

> we could (and we should) do the iteration [of the keys in a Secret] in lexicographical order

*[clarification is from me]* OK

> but given this is a generic high level document we can leave this as a design choice for each operator and document it in the related section

I disagree here. I think the exact meaning (i.e. the resulting oslo configuration) of the multi keys Secret case should be defined here.  We don't want two operators defining the same `customServiceConfigSecrets` field with the same syntaxt **but with different semantic**. As that would be highly confusing and error prone to operator by the end user.

Reply:
Yes, but I'm not sure we're defining how the operator should parse the list of secrets referenced via the `customServiceConfigSecrets` interface: I'm ok to propose to iterate in a given order (lexicographical) to avoid problems like the one described in the previous comments, I'm less sure about opening a `semantic` related conversation in this document.

Reply:
Done